packages/webview_flutter/android/src/main/java/io/flutter/plugins/webviewflutter/FlutterWebViewClient.java

```diff
@@ -0,0 +1,112 @@
+package io.flutter.plugins.webviewflutter;
+
+import android.annotation.TargetApi;
+import android.graphics.Bitmap;
+import android.os.Build;
+import android.util.Log;
+import android.webkit.WebResourceError;
+import android.webkit.WebResourceRequest;
+import android.webkit.WebView;
+import android.webkit.WebViewClient;
+import io.flutter.plugin.common.MethodChannel;
+import java.util.HashMap;
+import java.util.Map;
+
+public class FlutterWebViewClient extends WebViewClient {
+  private static final String TAG = FlutterWebViewClient.class.getSimpleName();
+
+  private final MethodChannel methodChannel;
+
+  private Boolean shouldOverrideUrlLoading = false;
+
+  FlutterWebViewClient(MethodChannel methodChannel) {
+    assert methodChannel != null;
+    this.methodChannel = methodChannel;
+  }
+
+  @Override
+  public boolean shouldOverrideUrlLoading(WebView view, String url) {
+    return onShouldOverrideUrlLoading(view, url);
+  }
+
+  @TargetApi(Build.VERSION_CODES.LOLLIPOP)
+  @Override
+  public boolean shouldOverrideUrlLoading(WebView view, WebResourceRequest request) {
+    return onShouldOverrideUrlLoading(view, request.getUrl().toString());
+  }
+
+  private boolean onShouldOverrideUrlLoading(final WebView view, final String url) {
+    if (shouldOverrideUrlLoading) {
+      Map<String, Object> args = new HashMap<>();
+      args.put("url", url);
+      methodChannel.invokeMethod(
+          "shouldOverrideUrlLoading",
+          args,
+          new MethodChannel.Result() {
+            @Override
+            public void success(Object result) {
+              shouldOverrideUrlLoading = (Boolean) result;
+              if (!shouldOverrideUrlLoading) {
+                view.loadUrl(url);
+              }
+            }
+
+            @Override
+            public void error(String errorCode, String errorMessage, Object errorDetails) {
+              Log.e(
+                  TAG,
+                  String.format(
+                      "Failed to handle channel reply: %s: %s" + errorCode, errorMessage));
+            }
+
+            @Override
+            public void notImplemented() {}
+          });
+      return true;
+    } else {
+      shouldOverrideUrlLoading = true;
```

Comment:
seems like this could be slightly wrong. suppose `shouldOverrideUrlLoading` is true for request 0, our Dart code gets called. Then the methodChannelResponse comes back with `false`. Request 1 happens, our Dart code doesn't get called and shouldOverrideUrlLoading gets set back to true. Request 2 happens and our Dart code gets called. If we don't want url overriding we'd end up getting Dart code called every other request I think.

Reply:
https://github.com/flutter/plugins/pull/1033/files#diff-7ea9273e5046b42ba55a038139bb843cR65 is trying to do something similar here but using a regex rather than a boolean to determine what to do. might be a bit more flexible although I don't like how they named it.

Reply:
Thank you for the review.
https://github.com/flutter/plugins/pull/996#discussion_r245469821
The issue is fixed https://github.com/flutter/plugins/pull/996/commits/3720d96bb136a80eb46a17fd1f1eb8887cb32e9c .